linkerd/transport-metrics/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+#![deny(warnings, rust_2018_idioms)]
+#![forbid(unsafe_code)]
+
+mod client;
+mod report;
+mod sensor;
+mod server;
+
+pub use self::{
+    client::Client,
+    report::Report,
+    sensor::{Sensor, SensorIo},
+    server::NewServer,
+};
+use linkerd_errno::Errno;
+use linkerd_metrics::{metrics, Counter, FmtLabels, Gauge, LastUpdate, Store};
+use parking_lot::Mutex;
+use std::{
+    collections::HashMap,
+    fmt,
+    hash::Hash,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+metrics! {
+    tcp_open_total: Counter { "Total count of opened connections" },
+    tcp_open_connections: Gauge { "Number of currently-open connections" },
+    tcp_read_bytes_total: Counter { "Total count of bytes read from peers" },
+    tcp_write_bytes_total: Counter { "Total count of bytes written to peers" },
+
+    tcp_close_total: Counter { "Total count of closed connections" }
+}
+
+pub fn new<K: Eq + Hash + FmtLabels>(retain_idle: Duration) -> (Registry<K>, Report<K>) {
+    let inner = Arc::new(Mutex::new(Inner::new()));
+    let report = Report::new(inner.clone(), retain_idle);
+    (Registry(inner), report)
+}
+
+#[derive(Clone, Debug)]
+pub struct Registry<K: Eq + Hash + FmtLabels>(Arc<Mutex<Inner<K>>>);
+
+type Inner<K> = Store<K, Metrics>;
+
+/// Stores a class of transport's metrics.
+#[derive(Debug, Default)]
+pub struct Metrics {
+    open_total: Counter,
+    open_connections: Gauge,
+    write_bytes_total: Counter,
+    read_bytes_total: Counter,
+
+    by_eos: Arc<Mutex<ByEos>>,
+}
+
+#[derive(Debug)]
+struct ByEos {
+    last_update: Instant,
+    metrics: HashMap<Eos, EosMetrics>,
+}
+
+/// Describes a classtransport end.
```

Comment:
typo (looks like it was copied and pasted from the old code though):
```suggestion
/// Describes a class of transport end.
```